verifier/verifier.go

```diff
@@ -113,20 +134,35 @@ func (v *Verifier) convertOptions(opt *VerificationOptions) (out x509.VerifyOpti
 // checks if the Name in the VerificationOptions matches the name on the
 // certificate.
 func (v *Verifier) Verify(c *x509.Certificate, opts VerificationOptions) (res *VerificationResult) {
+	opts.clean()
+	xopts := v.convertOptions(&opts)
+
 	res = new(VerificationResult)
 	res.Name = opts.Name
+	res.InValidityWindow = c.TimeInValidityPeriod(opts.VerifyTime)
 
-	if res.ValidationError == nil {
-		xopts := v.convertOptions(&opts)
+	// Don't pass DNSName to x509.Verify(), we'll check it ourselves by calling
+	// VerifyHostname() if necessary.
+	xopts.DNSName = ""
 
-		// Don't pass DNSName to x509.Verify(), we'll check it ourselves by calling
-		// VerifyHostname() if necessary.
-		xopts.DNSName = ""
+	res.Current, res.Expired, res.Never, res.ValidationError = c.Verify(xopts)
+	if len(opts.Name) > 0 {
+		res.NameError = c.VerifyHostname(opts.Name)
+	}
 
-		res.Current, res.Expired, res.Never, res.ValidationError = c.Verify(xopts)
-		if len(opts.Name) > 0 {
-			res.NameError = c.VerifyHostname(opts.Name)
-		}
+	// Determine certificate type.
+	if xopts.Roots.Contains(c) {
+		// A certificate is only a root if it's in the root store.
+		res.CertificateType = x509.CertificateTypeRoot
+	} else if c.IsCA && !c.SelfSigned {
```

Comment:
C1 has K1 , subj=S1, issuer=I1, S1==I1, sn=SN1, is a root
C2 has K1, S1, I1, sn=SN2, is not a root. 
C3 is signed by C2. Could not have been signed by C1 because AKID includes C2 SN.